src/drivers/redis/Queue.php

```diff
@@ -56,13 +73,26 @@ public function run($repeat, $timeout = 0)
     {
         return $this->runWorker(function (callable $canContinue) use ($repeat, $timeout) {
             while ($canContinue()) {
-                if (($payload = $this->reserve($timeout)) !== null) {
-                    list($id, $message, $ttr, $attempt) = $payload;
-                    if ($this->handleMessage($id, $message, $ttr, $attempt)) {
-                        $this->delete($id);
+                if ($this->acquire()) {
+                    try {
+                        $payload = $this->reserve($timeout);
+                    } finally {
+                        $this->release();
+                    }
+
+                    if ($payload !== null) {
+                        list($id, $message, $ttr, $attempt) = $payload;
+                        if ($this->handleMessage($id, $message, $ttr, $attempt)) {
+                            $this->delete($id);
+                        } else {
+                            // job is failed but we want to return it back into
+                            // the queue
+                            $this->redis->zadd("$this->channel.reserved", time(), $id);
```

Comment:
That cold lead to fail-add indefinite cycle. Should be limited number of re-adding the job back.

Reply:
Yep, you are right, but I rely on a class/global defined number of attempts. In this case if all attempts have been used up, job would be marked as a handled and this condition will never be invoked.